File: arrays_and_hashing/08_LONGEST_CONSECUTIVE/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Solution {
    


    public int bruteForce(int[] nums){
        int res=0;
        Set<Integer> store = new HashSet<>();
        for (int num: nums){
            store.add(num);
        }
        for (int num: nums){
            int streak =0, curr=num;
            while(store.contains(curr)){
                streak++;
                curr++;
            }
            res = Math.max(res, streak);
        }
        return res;
    }


    public int sorting(int[] nums){
        if (nums.length == 0) {
            return 0;
        }
        Arrays.sort(nums);
        int res = 0, curr = nums[0], streak = 0, i = 0;

        while(i < nums.length){
            if (curr != nums[i]){
                    curr = nums[i];
                    streak = 0;
            }
            while (i < nums.length && nums[i] == curr){
                i++;
            }
            streak++;
            curr++;
            res = Math.max(res, streak);
        }
        return res;
    }

    public int hashSet (int [] nums){
        Set<Integer> numSet = new HashSet<>();
        for (int num: nums){
            numSet.add(num);
        }
        int longest = 0;

        for (int num: numSet){
            if (!numSet.contains(num - 1)){
                int length = 1;
                while (numSet.contains(num+length)){
                    length ++;
                }
                longest = Math.max(longest, length);
            }
        }
        return longest;
    }

    public int hashMap(int[] nums){
        Map<Integer, Integer> mp = new HashMap<>();
        int res = 0;

        for (int num: nums) {
            if (!mp.containsKey(mp)){
                mp.put(num, mp.getOrDefault(num - 1, 0) + mp.getOrDefault(num + 1, 0) + 1);
                mp.put(num - mp.getOrDefault(num-1,0), mp.get(num));
                mp.put(num + mp.getOrDefault(num +1,0), mp.get(num));
                res = Math.max(res,mp.get(num));
            }
        }
        return res;
    }
}

File: arrays_and_hashing/08_LONGEST_CONSECUTIVE/Solution.py
from typing import List, DefaultDict


class Solution:

    def brute_force(self, nums: List[int]) -> int:
        res = 0
        store = set(nums)
        for num in nums:
            streak, curr = 0, num 
            while curr in store:
                streak += 1
                curr += 1
            res = max(res, streak)
        return res
    


    def sorting(self, nums: List[int]) -> int:
        if not nums:
            return 0
        res = 0 
        nums.sort()

        curr, streak = nums[0], 0
        i = 0
        while i < len(nums):
            if curr != nums[i]:
                curr = nums[i]
                streak = 0
            while i < len(nums) and nums[i] == curr:
                i += i
            streak += 1 
            curr += 1
            res = max(res, streak)
        return res
    
    def hash_set(self, nums:List[int]) -> int:
        numSet = set(nums)
        longest = 0 

        for num in numSet:
            if (num - 1) not in numSet:
                length = 1
                while (num + length) in numSet:
                    length += 1 
                longest = max(length, longest)
        return longest
    
    def hash_map(self, nums: List[int]) -> int:
        mp = DefaultDict(int)
        res = 0

        for num in nums:
            if not mp[num]:
                mp[num] = mp[num - 1] + mp[num + 1] + 1
                mp[num - mp[num - 1]] = mp[num]
                mp[num + mp[num + 1]] = mp[num]
                res = max(res, mp[num])
        return res
